lib/protocol/protocol_test.go

```diff
@@ -125,6 +125,53 @@ func TestCloseOnBlockingSend(t *testing.T) {
 	}
 }
 
+func TestCloseRace(t *testing.T) {
+	indexReceived := make(chan struct{})
+	unblockIndex := make(chan struct{})
+	m0 := newTestModel()
+	m0.indexFn = func(_ DeviceID, _ string, _ []FileInfo) {
+		close(indexReceived)
+		<-unblockIndex
+	}
+	m1 := newTestModel()
+
+	ar, aw := io.Pipe()
+	br, bw := io.Pipe()
+
+	c0 := NewConnection(c0ID, ar, bw, m0, "c0", CompressNever).(wireFormatConnection).Connection.(*rawConnection)
+	c0.Start()
+	c1 := NewConnection(c1ID, br, aw, m1, "c1", CompressNever)
+	c1.Start()
+	c0.ClusterConfig(ClusterConfig{})
+	c1.ClusterConfig(ClusterConfig{})
+
+	c1.Index("default", nil)
+	select {
+	case <-indexReceived:
+	case <-time.After(time.Second):
+		t.Fatal("timed out before receiving index")
+	}
+
+	go c0.internalClose(errManual)
+	select {
+	case <-c0.closed:
+	case <-time.After(time.Second):
+		t.Fatal("timed out before c0.closed was closed")
+	}
+
+	select {
+	case <-m0.closedCh:
+		t.Errorf("receiver.Closed called before receiver.Index")
+	default:
```

Comment:
This is somewhat racey between the internalClose doing it's thing and actually getting to this point to check.
For all that we know interalClose hasn't even run yet, hence this channel is not closed.

Reply:
It should be, but the chain of events is indeed a bit obscure:
1. Sending index, which blocks the `receiver-/dispatcherloop`.
2. Calling `internalClose` in a goroutine: Closes `c0.closed`, then blocks on waiting for the `receiver-/dispatcherloop` to terminate.
3. Check that `c0` is closed and `m0` did not yet receive a call to `Closed`.
4. Unblock sending index, which unblocks and then lets the `receiver-/dispatcherloop` terminate.
5. Check that now `m0` got `Closed`.